File: app/src/main/java/com/elementary/tasks/navigation/settings/additional/TemplateActivity.java
package com.elementary.tasks.navigation.settings.additional;

import android.content.ContentResolver;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.net.Uri;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import com.elementary.tasks.R;
import com.elementary.tasks.core.ThemedActivity;
import com.elementary.tasks.core.utils.BackupTool;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.utils.RealmDb;
import com.elementary.tasks.core.utils.TimeUtil;
import com.elementary.tasks.databinding.ActivityTemplateLayoutBinding;

import java.io.IOException;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class TemplateActivity extends ThemedActivity {

    private static final int MENU_ITEM_DELETE = 12;

    private ActivityTemplateLayoutBinding binding;
    private TemplateItem mItem;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        loadTemplate();
        binding = DataBindingUtil.setContentView(this, R.layout.activity_template_layout);
        initActionBar();
        initMessageField();
        showTemplate();
    }

    private void loadTemplate() {
        Intent intent = getIntent();
        String id = intent.getStringExtra(Constants.INTENT_ID);
        if (id != null) {
            mItem = RealmDb.getInstance().getTemplate(id);
        } else if (intent.getData() != null) {
            try {
                Uri name = intent.getData();
                String scheme = name.getScheme();
                if (ContentResolver.SCHEME_CONTENT.equals(scheme)) {
                    ContentResolver cr = getContentResolver();
                    mItem = BackupTool.getInstance().getTemplate(cr, name);
                } else {
                    mItem = BackupTool.getInstance().getTemplate(name.getPath(), null);
                }
            } catch (IOException | IllegalStateException e) {
                e.printStackTrace();
            }
        }
    }

    private void initMessageField() {
        binding.messageInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                updateLeftView(s.length());
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    private void updateLeftView(int count) {
        binding.leftCharacters.setText(String.format(getString(R.string.left_characters_x), (120 - count) + ""));
    }

    private void showTemplate() {
        if (mItem != null) {
            binding.messageInput.setText(mItem.getTitle());
            String title = mItem.getTitle();
            if (title != null) {
                updateLeftView(title.length());
            }
        }
    }

    private void initActionBar() {
        setSupportActionBar(binding.toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayShowTitleEnabled(false);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
        }
        binding.toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24dp);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_add:
                saveTemplate();
                return true;
            case android.R.id.home:
                finish();
                return true;
            case MENU_ITEM_DELETE:
                deleteItem();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void deleteItem() {
        if (mItem != null) {
            RealmDb.getInstance().deleteTemplates(mItem);
            new DeleteTemplateFilesAsync(this).execute(mItem.getKey());
        }
        finish();
    }

    private void saveTemplate() {
        String text = binding.messageInput.getText().toString().trim();
        if (text.length() == 0) {
            binding.messageInput.setError(getString(R.string.must_be_not_empty));
            return;
        }
        String date = TimeUtil.getGmtDateTime();
        if (mItem != null){
            mItem.setDate(date);
            mItem.setTitle(text);
        } else {
            mItem = new TemplateItem(text, date);
        }
        RealmDb.getInstance().saveObject(mItem);
        finish();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_create_template, menu);
        if (mItem != null) {
            menu.add(Menu.NONE, MENU_ITEM_DELETE, 100, getString(R.string.delete));
        }
        return true;
    }
}

File: app/src/main/java/com/elementary/tasks/navigation/settings/additional/TemplateActivity.kt
package com.elementary.tasks.navigation.settings.additional

import android.content.ContentResolver
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.elementary.tasks.R
import com.elementary.tasks.core.arch.BindingActivity
import com.elementary.tasks.core.data.models.SmsTemplate
import com.elementary.tasks.core.utils.*
import com.elementary.tasks.core.view_models.Commands
import com.elementary.tasks.core.view_models.sms_templates.SmsTemplateViewModel
import com.elementary.tasks.databinding.ActivityTemplateBinding
import java.util.*

class TemplateActivity : BindingActivity<ActivityTemplateBinding>(R.layout.activity_template) {

    private val viewModel: SmsTemplateViewModel by lazy {
        ViewModelProviders.of(this, SmsTemplateViewModel.Factory(getId()))
                .get(SmsTemplateViewModel::class.java)
    }
    private var mItem: SmsTemplate? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initActionBar()
        loadTemplate()
    }

    private fun getId(): String = intent.getStringExtra(Constants.INTENT_ID) ?: ""

    private fun loadTemplate() {
        val intent = intent
        initViewModel()
        if (intent.data != null) {
            readUri()
        } else if (intent.hasExtra(Constants.INTENT_ITEM)) {
            try {
                val item = intent.getParcelableExtra(Constants.INTENT_ITEM) as SmsTemplate?
                if (item != null) {
                    showTemplate(item, true)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun readUri() {
        if (!Permissions.checkPermission(this, SD_REQ, Permissions.READ_EXTERNAL)) {
            return
        }
        intent.data?.let {
            try {
                mItem = if (ContentResolver.SCHEME_CONTENT != it.scheme) {
                    val any = MemoryUtil.readFromUri(this, it)
                    if (any != null && any is SmsTemplate) {
                        any
                    } else null
                } else null
                mItem?.let { item -> showTemplate(item, true) }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun initViewModel() {
        viewModel.smsTemplate.observe(this, Observer{ smsTemplate ->
            if (smsTemplate != null) {
                showTemplate(smsTemplate)
            }
        })
        viewModel.result.observe(this, Observer{ commands ->
            if (commands != null) {
                when (commands) {
                    Commands.SAVED, Commands.DELETED -> finish()
                    else -> {
                    }
                }
            }
        })
    }

    private fun showTemplate(smsTemplate: SmsTemplate, fromFile: Boolean = false) {
        this.mItem = smsTemplate
        binding.toolbar.title = getString(R.string.edit_template)
        if (!viewModel.isEdited) {
            binding.messageInput.setText(smsTemplate.title)
            viewModel.isEdited = true
            viewModel.isFromFile = fromFile
            if (fromFile) {
                viewModel.findSame(smsTemplate.key)
            }
        }
    }

    private fun initActionBar() {
        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeButtonEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(true)
        binding.toolbar.navigationIcon = ViewUtils.backIcon(this, isDarkMode)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_add -> {
                askCopySaving()
                return true
            }
            android.R.id.home -> {
                finish()
                return true
            }
            MENU_ITEM_DELETE -> {
                deleteItem()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun askCopySaving() {
        if (viewModel.isFromFile && viewModel.hasSameInDb) {
            dialogues.getMaterialDialog(this)
                    .setMessage(R.string.same_template_message)
                    .setPositiveButton(R.string.keep) { dialogInterface, _ ->
                        dialogInterface.dismiss()
                        saveTemplate(true)
                    }
                    .setNegativeButton(R.string.replace) { dialogInterface, _ ->
                        dialogInterface.dismiss()
                        saveTemplate()
                    }
                    .setNeutralButton(R.string.cancel) { dialogInterface, _ ->
                        dialogInterface.dismiss()
                    }
                    .create()
                    .show()
        } else {
            saveTemplate()
        }
    }

    private fun deleteItem() {
        mItem?.let { viewModel.deleteSmsTemplate(it) }
    }

    private fun saveTemplate(newId: Boolean = false) {
        val text = binding.messageInput.text.toString().trim()
        if (text.isEmpty()) {
            binding.messageLayout.error = getString(R.string.must_be_not_empty)
            binding.messageLayout.isErrorEnabled = true
            return
        }
        val date = TimeUtil.gmtDateTime
        val item = (mItem ?: SmsTemplate()).apply {
            this.date = date
            this.title = text
        }
        if (newId) {
            item.key = UUID.randomUUID().toString()
        }
        viewModel.saveTemplate(item)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_create_template, menu)
        if (mItem != null && !viewModel.isFromFile) {
            menu.add(Menu.NONE, MENU_ITEM_DELETE, 100, getString(R.string.delete))
        }
        return true
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == SD_REQ && Permissions.checkPermission(grantResults)) {
            readUri()
        }
    }

    companion object {
        private const val MENU_ITEM_DELETE = 12
        private const val SD_REQ = 555
    }
}
